rust/template/tcp_channel/receiver.rs

```diff
@@ -25,15 +27,55 @@ use serde::de::DeserializeOwned;
 
 use crate::message::Message;
 
+#[derive(Debug)]
+enum Fd {
+    /// We are still listening for an incoming connection and
+    /// this is the corresponding file descriptor.
+    Listening(RawFd),
+    /// We have accepted a connection and read data from it.
+    Accepted(RawFd),
+}
+
+impl Fd {
+    fn close(&self) -> Result<(), Error> {
+        match self {
+            Fd::Listening(fd) | Fd::Accepted(fd) => {
+                let rc = unsafe { shutdown(*fd, SHUT_RDWR) };
+                if rc != 0 {
+                    return Err(Error::last_os_error());
+                }
+
+                let rc = unsafe { close(*fd) };
+                if rc != 0 {
+                    return Err(Error::last_os_error());
+                }
+                Ok(())
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+struct Data {
```

Comment:
`Data` is not a great name for this data structure. `ConnectionState` maybe?
Also, can't you just add the third option to `Fd` for closed state and not need this data type at all?

Reply:
Sure, I'll change the name. Yes, a third state may be possible. I'll try it out. I iterated through so many variants that when I finally had something working I did not think too hard to reduce the state even more.